biz.aQute.repository/src/aQute/bnd/repository/p2/provider/P2Indexer.java

```diff
@@ -131,10 +138,29 @@ private Repository readRepository() throws Exception {
 		List<Artifact> artifacts = p2.getArtifacts();
 		List<Promise<Resource>> fetched = new ArrayList<>(artifacts.size());
 		Set<URI> visited = new HashSet<>(artifacts.size());
+		Set<String> visitedHash = new HashSet<>(artifacts.size());
+		Map<String,Resource> existing = new HashMap<>();
+
+		if (getBridge() != null) {
+			Requirement requirement = new RequirementBuilder(P2_CAPABILITY_NAMESPACE).addAttribute(MD5_ATTRIBUTE, "*")
+					.buildSyntheticRequirement();
+			for (Capability capability : getBridge().getRepository().findProviders(asList(requirement)).get(
+					requirement)) {
+				Resource resource = capability.getResource();
+				String md5 = (String) capability.getAttributes().get(MD5_ATTRIBUTE);
+				existing.put(md5, resource);
```

Comment:
I am concerned that MD5 alone is not strong enough be the sole key. The `existing` map here is meant to be a map from some unique bundle key to the unique bundle resource. Since MD5 is known to be weak, perhaps the key here should be composed of the bsn+version+md5. The key is the bundle as uniquely identified by its bsn and version and also a hash to detect changes.

This comment also applies to the `visitedHash` set. I think the current `visited` set should be renamed `visitedURI` and the new set `visitedArtifact`. The `existing` name is also not descriptive enough. I think a better name is `knownResources`?